Block/Service/AbstractCategoriesBlockService.php

```diff
@@ -0,0 +1,160 @@
+        $context  = $blockContext->getSetting('context');
+        $root     = $this->getCategoryManager()->getRootCategory($context);
+
+        return $this->renderResponse($blockContext->getTemplate(), array(
+            'context'  => $blockContext,
+            'settings' => $blockContext->getSettings(),
+            'block'    => $blockContext->getBlock(),
+            'category' => $category,
+            'root'     => $root,
+        ), $response);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildEditForm(FormMapper $formMapper, BlockInterface $block)
+    {
+        $context    = $block->getSetting('context');
+
+        $adminField = $this->getFormAdminType($formMapper, $this->getCategoryAdmin(), 'categoryId', 'category', array(
+            'label'              => 'form.label_category',
+        ), array(
+            'translation_domain' => 'SonataClassificationBundle',
+            'link_parameters'    => array(
+                array(
+                    'context' => $context,
+                ),
+            ),
+        ));
+
+        $formMapper->add('settings', 'sonata_type_immutable_array', array(
+            'keys' => array(
+                array('title', 'text', array(
+                    'label'    => 'form.label_title',
+                    'required' => false,
+                )),
+                array('context', 'choice', array(
+                    'label'    => 'form.label_context',
+                    'required' => false,
+                    'choices'  => $this->getContextChoices(),
+                )),
+                array($adminField, null, array(
+                )),
+            ),
+            'translation_domain' => 'SonataClassificationBundle',
+        ));
+    }
+    /**
+     * {@inheritdoc}
+     */
+    public function configureSettings(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults(array(
+            'title'      => 'Categories',
+            'category'   => false,
+            'categoryId' => null,
+            'context'    => 'default',
+            'template'   => 'SonataClassificationBundle:Block:base_block_categories.html.twig',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function load(BlockInterface $block)
+    {
+        if (is_numeric($block->getSetting('categoryId', null))) {
+            $block->setSetting('categoryId', $this->getCategory($block->getSetting('categoryId', null)));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function prePersist(BlockInterface $block)
+    {
+        $id = is_object($block->getSetting('categoryId')) ? $block->getSetting('categoryId')->getId() : null;
+
+        $block->setSetting('categoryId', $id);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function preUpdate(BlockInterface $block)
+    {
+        $id = is_object($block->getSetting('categoryId')) ? $block->getSetting('categoryId')->getId() : null;
+
```

Comment:
Thanks for changing that : please note that OTVs can still be OK in some cases : https://a-nickels-worth.blogspot.fr/2016/04/a-guide-to-naming-variables.html